cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/ProgramStatusEventPublisher.java

```diff
@@ -163,41 +166,54 @@ protected void processMessages(StructuredTableContext structuredTableContext,
         .withUserArgs(GSON.fromJson(userArgsString, argsMapType))
         .withSystemArgs(GSON.fromJson(sysArgsString, argsMapType));
       if (programRunStatus.isEndState()) {
-        populateErrorDetailsAndMetrics(builder, properties, programRunStatus, programRunId, publishTime);
+        populateErrorDetailsAndMetrics(builder, properties, programRunStatus, programRunId, notifiedTime);
       } else {
-        writeInEventWriters(builder, publishTime);
+        writeInEventWriters(builder, notifiedTime);
       }
     });
   }
 
-  private void writeInEventWriters(ProgramStatusEventDetails.Builder builder, long publishTime) {
+  private void writeInEventWriters(ProgramStatusEventDetails.Builder builder, long notifiedTime) {
+    long publishTime = System.currentTimeMillis();
     ProgramStatusEventDetails programStatusEventDetails = builder.build();
     ProgramStatusEvent programStatusEvent = new ProgramStatusEvent(publishTime, EVENT_VERSION,
                                                                    instanceName,
                                                                    projectName, programStatusEventDetails);
     List<ProgramStatusEvent> listProgramStatus = new ArrayList<>();
     listProgramStatus.add(programStatusEvent);
+    emitMetrics(notifiedTime, publishTime, programStatusEvent);
     this.eventWriters.forEach(eventWriter -> eventWriter.write(listProgramStatus));
   }
 
+  private void emitMetrics(long notifiedTime, long publishTime, ProgramStatusEvent programStatusEvent) {
+    Map<String, String> metricTags = new HashMap<>();
+    metricTags.put(Constants.Metrics.Tag.NAMESPACE, programStatusEvent.getEventDetails().getNamespace());
+    metricTags.put(Constants.Metrics.Tag.STATUS, programStatusEvent.getEventDetails().getStatus());
+    metricTags.put(Constants.Metrics.Tag.PROGRAM, programStatusEvent.getEventDetails().getProgramName());
+    metricTags.put(Constants.Metrics.Tag.APP, programStatusEvent.getEventDetails().getApplicationName());
+    metricsCollectionService.getContext(metricTags).increment(Constants.Metrics.ProgramEvent.PUBLISHED_COUNT, 1L);
+    metricsCollectionService.getContext(metricTags)
+      .gauge(Constants.Metrics.ProgramEvent.PUBLISH_LATENCY_MS, publishTime - notifiedTime);
```

Comment:
wouldnt `publishTime ` and `notifiedTime ` be few millis apart if not less? Wondering what latency would it track?

Reply:
This will be relevant if there is a significant delay trying to fetch program metrics from spark history server - https://github.com/cdapio/cdap/blob/7f899d6bb4d7cd17b6f55e369832541eb72965f8/cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/ProgramStatusEventPublisher.java#L197

Reply:
I see.

Reply:
On further thinking and discussion, decided to move the latency metrics to the SparkProgramStatusMetricsProvider where this will make more sense.